doc/devguide/extending/app-layer/transactions.rst

```diff
@@ -0,0 +1,299 @@
+************
+Transactions
+************
+
+.. contents:: Table of Contents
+
+_`General Concepts`
+===================
+
+Transactions are abstractions that help detecting and logging in Suricata. They also help during the detection phase,
+when dealing with protocols that can have large PDUs, like TCP, in controlling state for partial rule matching, in case of rules that mention more than one field.
+
+Transactions are implemented and stored in the per-flow state. The engine interacts with them using a set of callbacks the parser registers.
+
+_`How the engine uses transactions`
+===================================
+
+Logging
+~~~~~~~
+
+Suricata controls when logging should happen based on transaction completeness. For simpler protocols, such as ``dns``
+or ``ntp``, that will most
+likely happen once per transaction, by the time of its completion. In other cases, like with HTTP, this may happen at intermediary states.
+
+In ``OutputTxLog``, the engine will compare current state with the value defined for the logging to happen, per flow
+direction (``logger->tc_log_progress``, ``logger->ts_log_progress``). If state is less than that value, the engine skips to
+the next logger. Code snippet from: suricata/src/output-tx.c:
+
+.. code-block:: c
+
+    static TmEcode OutputTxLog(ThreadVars *tv, Packet *p, void *thread_data)
+    {
+        .
+        .
+        .
+            if ((ts_eof && tc_eof) || last_pseudo) {
+                SCLogDebug("EOF, so log now");
+            } else {
+                if (logger->LogCondition) {
+                    int r = logger->LogCondition(tv, p, alstate, tx, tx_id);
+                    if (r == FALSE) {
+                        SCLogDebug("conditions not met, not logging");
+                        goto next_logger;
+                    }
+                } else {
+                    if (tx_progress_tc < logger->tc_log_progress) {
+                        SCLogDebug("progress not far enough, not logging");
+                        goto next_logger;
+                    }
+
+                    if (tx_progress_ts < logger->ts_log_progress) {
+                        SCLogDebug("progress not far enough, not logging");
+                        goto next_logger;
+                    }
+                }
+             }
+        .
+        .
+        .
+    }
+
+Rule Matching
+~~~~~~~~~~~~~
+
+Transaction progress is also used for certain keywords to know what is the minimum state before we can expect a match: until that, Suricata won't even try to look for the patterns.
```

Comment:
Would it be nice to have the list of such keywords when running `./src/suricata --list-keywords=all`

Reply:
Could it be like a linked document, which I'd mention here?

Reply:
I would rather like something automatically generated...

Reply:
Ok, could we generate that list automatically, but have the list on a different page, to prevent this one from getting too long?

Reply:
Yes, but my point was that this requires another development work first.
I do not know how to get that list

Reply:
Oh, ok, sorry, I had gotten it wrongly at first. So we need something that allows us to list what keywords need that minimum stage, before matching? Should I create a ticket for that?

Reply:
> So we need something that allows us to list what keywords need that minimum stage, before matching?

I would first like to have the list of keywords that are matched within a transaction (and the what the other ones are matched on)

Reply:
Ok, I think I must dig more in order to get a better grasp at this...